### PR TITLE
Migrate to serverspec tests

### PR DESCRIPTION
**Pull request by patcon**
The current minitest chef_handler stuff seems to be out-of-vogue. Normally, this wouldn't really matter, but given that it required chef_gem resources and compile time activity (which is now considered something to be wary of and uses minimally), I think we'd do well to move to busser-serverspec.

I've got most of the work done on a branch, but haven't create a PR yet. If I submit a minimally intrusive PR (with no new tests to begin with), would that be acceptable?

### CODE REVIEW

**Comment by patcon** [comment]
Heh. Took less effort than expected. I'll get the postgresql tests going in another PR :)

To be fair, I'm still running test-kitchen on the platforms, so probably good to wait on that first :)

File: Berksfile [hunk 1]
```diff
@@ -3,7 +3,12 @@ source 'https://supermarket.chef.io'
 metadata
 
 group :integration do
-  cookbook 'java'
-  cookbook 'minitest-handler'
-  cookbook 'stash_test', :path => 'test/cookbooks/stash_test'
+  cookbook 'java',
+    # Pending release: ark_java fails when tar is not installed
+    # See: https://github.com/agileorbit-cookbooks/java/issues/263
+    github: 'agileorbit-cookbooks/java'
```
Comment (patcon) [inline Berksfile]: We've got some pointers in the Berksfile to unpinned git repos, so I'm wondering whether we should commit the Berksfile lock to eliminate noise from the odd broken commit... Thoughts?

Or we could pin to a specific commit in these repos. Your call

Comment (linc01n) [inline Berksfile]: My only concern of pinning the other cookbook will raise conflict.

e.g. java
Other cookbook may depend to an older version if we pin to the github HEAD. 

And if we pinned to a specific version,  Berkshelf may not handle the dependency resolution well.

Comment (patcon) [inline Berksfile]: As I understand it, cookbooks that depend on this one never ever look at the Berksfile and its lockfile when doing dependency resolution -- just the metadata.rb. These are just used for us :)

The downside of locking is that our travis tests (once we get them going) might not immediately reveal a failure that I user might be seeing when an upstream cookbook breaks something -- we'd have to explicitly update the lockfile and commit it to see that failure. Either one of these approached could work for us

Comment (patcon) [inline Berksfile]: Yep, seems some cookbooks even add them the chefignore:
https://github.com/ramonskie/bamboo/blob/master/chefignore#L72

**Comment by patcon** [comment]
Platforms that I've confirmed working:

- [x] centos-6
- [ ] centos-7
- [ ] ubuntu-12.04
- [x] ubuntu-14.04

The boxes are taking awhile to download for the remaining platforms, so I'll have to leave that for later. Feel free to test those yourself if you have time, but otherwise I'll confirm tomorrow

**Comment by patcon** [comment]
Good to go!

**Comment by linc01n** [comment]
:+1: Merged

**Comment by legal90** [comment]
@patcon Could you please explain why have you added `test-helper` dependency to the `Berksfile`?
As I know, chef-zero provisioner for TestKitchen automatically dumps the result node object to JSON: `/tmp/kitchen/nodes/<instance>.json` (on the node itself). So, what is the benefit of this helper?

**Comment by patcon** [comment]
Good point. I think I started using this pattern before chef_zero provisioner was default, and just never re-thought it :)

I suspect it's not needed, but I opened an issue in their queue for feedback (auto-ref'd above)

**Comment by patcon** [comment]
hm. As was brought up in the linked issue, the node object isn't a deepmerged hash, so it would seem that in order to use that, chef would need to be installed in the[ busser-serverspec gemspec](https://github.com/test-kitchen/busser-serverspec/blob/master/busser-serverspec.gemspec)

So it seems the cookbook is still useful. But I could be wrong

**Comment by patcon** [comment]
Or wait, we'd need a custom gemfile with chef:
https://github.com/test-kitchen/busser-serverspec#-usage

All-in-all, I'd almost rather use the test-helper cookbook, but open to other options if someone wants to investigate

**Comment by legal90** [comment]
@patcon Ah, now I got how it works. You dump the node object to the JSON file on the host machine in order to use this JSON in integration tests on "verify" step.
https://github.com/bflad/chef-stash/blob/82fa40956d4116866b28368741dea4db7d77d5f7/test/integration/helpers/serverspec/spec_helper.rb#L6
Well, you can disregard my question. And thank you for explanation. :+1: 

P.s. But personally I don't like this approach. IMO, it's more reliable to use pre-defined values in tests rather than ones generated dynamically ([example in confluence cookbook](https://github.com/parallels-cookbooks/confluence/tree/master/test)).

**Comment by patcon** [comment]
fwiw, it's not always the right way, but when the logic is a little odd (particularly legacy anti-patterns like calculated attributes or state storing in attributes), sometimes you want it. We had a situation when setting a software version wasn't resulting in that version being installed, so it was very helpful to cross-check the `/bin/mytool --version` response with that specified in attr :)

**Pull request merged**